E-Freezer/freezer_sched_class.md

```diff
@@ -0,0 +1,704 @@
+invoked with a simple and well-defined API. The scheduler - from the perspective
+of the rest of the kernel - has two main responsibilities:
+
+* **Responsibility I**: Provide an interface to halt the currently running
+process and switch to a new one. To do so, the scheduler must pick the next
+process to run, which is a nontrivial problem.
+
+* **Responsibility II**: Indicate to the rest of the OS when a new process
+should be run.
+
+### Responsibility I: Switching to the Next Process
+
+To fulfill its first responsibility, the scheduler must somehow keep track of
+all the running processes.
+
+### The Runqueue
+
+Here's the first over-simplification: you can think of the scheduler as a system
+that maintains a simple queue of processes in the form of a linked list. The
+process at the head of the queue is allowed to run for some "time slice" - say,
+10 milliseconds. After this time slice expires, the process is moved to the back
+of the queue, and the next process gets to run on the CPU for the same time
+slice. When a running process is forcibly stopped and taken off the CPU in this
+way, we say that it has been **preempted**. The linked list of processes waiting
+to have a go on the CPU is called the runqueue. Each CPU has its own runqueue,
+and a given process may appear on only one CPU's runqueue at a time. Processes
+CAN migrate between various CPUs' runqueues, but we'll save this discussion for
+later.
+
+<div align='center'>
+    <img src='./final_simple_runqueue.png'/><br/>
+    Figure 1: An over-simplification of the runqueue
+</div>
+
+The scheduler is not really this simple; the runqueue is defined in the kernel
+as `struct rq`, and you can take a peek at its definition
+[here](https://elixir.bootlin.com/linux/v5.10.158/source/kernel/sched/sched.h#L897).
+Spoiler alert: it's not a linked list! To be fair, the explanation that I gave
+above more or less describes the very first Linux runqueue. But over the years,
+the scheduler evolved to incorporate multiple scheduling algorithms. These
+include:
+
+  * Completely Fair Scheduler (CFS)
+  * Real-Time Scheduler
+  * Deadline Scheduler
+
+The modern-day runqueue is no longer a linked list but actually a collection of
+algorithm-specific runqueues corresponding to the list above. Indeed,
+`struct rq` has the following members:
+
+```c
+struct cfs_rq cfs;  // CFS scheduler runqueue
+struct rt_rq rt;    // Real-time scheduler runqueue
+struct dl_rq dl;    // Deadline scheduler runqueue
+```
+
+Keep these details in the back of your mind so that you don't get bogged down.
+Remember: the goal here is to understand how the scheduler interoperates with
+the rest of the kernel. The main takeaway is that a process is allowed to run
+for some time, and when that time expires, it gets preempted so that the next
+process can run.
+
+### Preemption vs Yielding
+
+Preemption is not always the reason a process is taken off the CPU. For example,

```

Comment:
Do we want to add some examples about when would preemption occur? The "time slicing" situation from last section is just one case. I think maybe add something at the end of last section:
"Note that there are some other situations that would lead to preemption. For example, when an interrupt occurs, the CPU may be preempted to handle the interrupt. Real-time processes may have higher priority than other processes and may preempt lower-priority processes to ensure that they meet their deadlines. "